### PR TITLE
More fine-grained CDN sync

### PR DESCRIPTION
**Pull request by ljedrz**
The current future-stream-based approach is pretty concise, but there are some issues with it:
- adjusting its logic is tricky
- error-handling is inconvenient
- it's hard to keep track of the number of pending blocks and predict the buffering behavior in general
- the backoff logic is more complicated than it needs to be
- despite best efforts, a shutdown may still cause storage issues
- some downloads can be very slow due to competition for shared objects

In this PR I propose that we instead switch to a simpler approach where we spawn a background task responsible for making concurrent requests to the CDN, and we can control both the number of desired concurrent requests and the number of pending blocks that we keep in memory.

In my local runs, it performs at least as fast as the current code, while providing us with a much greater degree of control over and insight into the progress. In addition, despite many restarts I haven't run into any storage failures.

### CODE REVIEW

File: node/cdn/src/blocks.rs [hunk 1]
```diff
@@ -138,143 +146,192 @@ pub async fn load_blocks<N: Network>(
...
+                                debug!(
+                                    "Received {ctx} in {:.2?} ({} queued for insertion)",
+                                    request_time.elapsed(),
+                                    pending_blocks.len()
+                                );
+                                break;
+                            }
+                            Err(error) => {
+                                // Increment the attempt counter, and wait with a linear backoff.
+                                attempts += 1;
```
Comment (joske) [inline node/cdn/src/blocks.rs]: should we have some max attempts?

Comment (ljedrz) [reply node/cdn/src/blocks.rs]: happy to impose some limit; @howardwu would you like to suggest one?

**Comment by ljedrz** [comment]
Rebased and applied a bit of refactoring to make the changes easier to digest.

File: node/cdn/src/blocks.rs (hunk 1 above)
Comment (ljedrz) [reply node/cdn/src/blocks.rs]: I introduced a limit of 10 attempts for now; if it is breached, the sync gets aborted.

**Comment by ljedrz** [comment]
With https://github.com/AleoHQ/snarkVM/pull/2322 applied I've almost managed to sync to the tip; tomorrow I'll be able to confirm that I've tested all possible sync scenarios using this PR.

**Comment by ljedrz** [comment]
I was able to sync to the tip :+1:, and at no time during my syncing have I encountered any storage issues despite many restarts.

**Review by howardwu — APPROVED** [review]

**Comment by howardwu** [comment]
@ljedrz can you confirm that the interfaces do not have any changes for the block height?

I see that the tests needed to be updated to reflect new range changes. This will cause breakages downstream for the Explorer, faucet, and all developers using the CDN to date.

**Pull request merged**